main/lsp/requests/completion.cc

```diff
@@ -92,36 +94,100 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method) {
+    string_view preDefinition = sourceCode.substr(0, sourceCode.rfind('\n', beginIndex));
 
-    int lastNewlineLoc = preDefinition.rfind('\n');
-    // if there is no '\n' in preDefinition, we're at the top of the file.
-    if (lastNewlineLoc == preDefinition.npos) {
-        return nullopt;
-    }
-    auto prevLine = preDefinition.substr(lastNewlineLoc, preDefinition.size() - lastNewlineLoc);
-    if (prevLine.find('#') == prevLine.npos) {
+    // Get all the lines before it.
+    std::vector<string_view> all_lines = absl::StrSplit(preDefinition, '\n');
+
+    // if there are no lines before the method definition, we're at the top of the file.
+    if (all_lines.empty()) {
         return nullopt;
     }
 
-    string documentation = "";
-    // keep looking for previous newline locations, searching for lines with # in them.
-    while (prevLine.find('#') != prevLine.npos) {
-        documentation = absl::StrCat(prevLine.substr(prevLine.find('#') + 1, prevLine.size()), "\n", documentation);
-        int prevNewlineLoc = preDefinition.rfind('\n', lastNewlineLoc - 1);
-        // if there is no '\n', we're at the top of the file, so just return documentation.
-        if (prevNewlineLoc == preDefinition.npos) {
+    std::vector<string_view> documentation_lines;
+
+    // Iterate from the last line, to the first line
+    for (auto it = all_lines.rbegin(); it != all_lines.rend(); it++) {
+        string_view line = absl::StripAsciiWhitespace(*it);
+
+        // Short circuit when line is empty
+        if (line.empty()) {
+            break;
+        }
+
+        // Handle single-line sig block
+        else if (absl::StartsWith(line, "sig")) {
+            // Do nothing for a one-line sig block
+        }
+
+        // Handle multi-line sig block
+        else if (absl::StartsWith(line, "end")) {
+            // ASSUMPTION: We either hit the start of file, a `sig do` or an `end`
+            do {
+                it++;
+                line = absl::StripAsciiWhitespace(*it);
+            } while (it != all_lines.rend()
+                    // Start of sig block
+                    && !absl::StartsWith(line, "sig do")
+                    // Invalid end keyword
+                    && !absl::StartsWith(line, "end"));
+
+            // We have either
+            // 1) Reached the start of the file
+            // 2) Found a `sig do`
+            // 3) Found an invalid end keyword
+            if (it == all_lines.rend() || absl::StartsWith(line, "end")) {
+                break;
+            } else {
+                // Reached a sig block.
+                ENFORCE(absl::StartsWith(line, "sig do"));
+
+                // Stop looking if this is a single-line block e.g `sig do; <block>; end`
+                string_view after_sig_do = line.substr(strlen("sig do"));
+                after_sig_do = absl::StripAsciiWhitespace(after_sig_do);
+                if (absl::StartsWith(after_sig_do, ";") && absl::EndsWith(after_sig_do, "end")) {
```

Comment:
I know this is faster than `absl::StartsWith("sig do;")`, but it's messy. Let's just do `absl::StartsWith(line, "sig do;") && absl::EndsWith(line, "end")`.

Reply:
Heck, then you can just make this:

```c++
if (it==all_ends.rend() || absl::StartsWith(line, "end") || (absl::StartsWith(line, "sig do;") && absl::EndsWith(line, "end")) { break; }
```

Reply:
Sounds good. I kept it after the iteration to keep it cleaner 6b16107c6